From 6cc26777b049d5972ccc68bdab132be8b06cd85f Mon Sep 17 00:00:00 2001
From: Protonu Basu <protonu@fb.com>
Date: Tue, 24 Mar 2020 21:21:03 -0700
Subject: [PATCH] fixing build failures with USE_NATIVE_ARCH ON

---
 caffe2/CMakeLists.txt | 1 +
 1 file changed, 1 insertion(+)

diff --git a/caffe2/CMakeLists.txt b/caffe2/CMakeLists.txt
index 4f76cd76752c2..0c1cbb8b38cb9 100644
--- a/caffe2/CMakeLists.txt
+++ b/caffe2/CMakeLists.txt
@@ -1208,6 +1208,7 @@ if (BUILD_TEST)
     add_executable(${test_name} "${test_src}")
     target_link_libraries(${test_name} ${Caffe2_MAIN_LIBS} gtest_main)
     target_include_directories(${test_name} PRIVATE $<INSTALL_INTERFACE:include>)
+    target_include_directories(${test_name} PRIVATE $<BUILD_INTERFACE:${CMAKE_BINARY_DIR}/include>)
     target_include_directories(${test_name} PRIVATE ${Caffe2_CPU_INCLUDE})
     add_test(NAME ${test_name} COMMAND $<TARGET_FILE:${test_name}>)
     if (INSTALL_TEST)
