--- a/torch/lib/libshm/core.cpp	2020-06-02 11:06:37.289324818 +0800
+++ b/torch/lib/libshm/core.cpp	2020-06-02 11:06:37.289324818 +0800
@@ -32,7 +32,7 @@
     SYSCHECK_ERR_RETURN_NEG1(close(pipe_ends[0]));
     SYSCHECK_ERR_RETURN_NEG1(dup2(pipe_ends[1], 1)); // Replace stdout
     SYSCHECK_ERR_RETURN_NEG1(close(pipe_ends[1]));
-    execl(manager_executable_path.c_str(), "torch_shm_manager", NULL);
+    execl(manager_executable_path.c_str(), "torch_shm_manager", (char*)0);
     exit(1);
   }
   SYSCHECK_ERR_RETURN_NEG1(close(pipe_ends[1]));
