--- ipc/chromium/src/third_party/libevent/CMakeLists.txt
+++ ipc/chromium/src/third_party/libevent/CMakeLists.txt
@@ -295,6 +295,7 @@ if (${GNUC})
 
     list(APPEND __FLAGS
          -Wall -Wextra -Wno-unused-parameter -Wstrict-aliasing -Wstrict-prototypes
+         -Wundef
 
          -fno-strict-aliasing # gcc 2.9.5+
          -Wmissing-prototypes
--- ipc/chromium/src/third_party/libevent/event-config.h.cmake
+++ ipc/chromium/src/third_party/libevent/event-config.h.cmake
@@ -485,6 +485,9 @@
 /* The size of 'void *', as computer by sizeof */
 #define EVENT__SIZEOF_VOID_P @EVENT__SIZEOF_VOID_P@
 
+/* The size of 'time_t', as computer by sizeof */
+#define EVENT__SIZEOF_TIME_T @EVENT__SIZEOF_TIME_T@
+
 /* Define to `__inline__' or `__inline' if that's what the C compiler
    calls it, or to nothing if 'inline' is not supported under any name.  */
 #ifndef __cplusplus
