--- widget/gtk/nsDeviceContextSpecG.cpp	2021-03-11 03:49:55.000000000 +0800
+++ widget/gtk/nsDeviceContextSpecG.cpp	2021-03-11 03:49:55.000000000 +0800
@@ -6,7 +6,6 @@
 #include "nsDeviceContextSpecG.h"
 
 #include "mozilla/gfx/PrintTargetPDF.h"
-#include "mozilla/gfx/PrintTargetPS.h"
 #include "mozilla/Logging.h"
 #include "mozilla/Services.h"
 
@@ -44,7 +43,6 @@
 using mozilla::gfx::IntSize;
 using mozilla::gfx::PrintTarget;
 using mozilla::gfx::PrintTargetPDF;
-using mozilla::gfx::PrintTargetPS;
 
 nsDeviceContextSpecGTK::nsDeviceContextSpecGTK()
     : mGtkPrintSettings(nullptr), mGtkPageSetup(nullptr) {}
@@ -117,12 +115,7 @@
     return PrintTargetPDF::CreateOrNull(stream, size);
   }
 
-  int32_t orientation = mPrintSettings->GetSheetOrientation();
-  return PrintTargetPS::CreateOrNull(
-      stream, size,
-      orientation == nsIPrintSettings::kPortraitOrientation
-          ? PrintTargetPS::PORTRAIT
-          : PrintTargetPS::LANDSCAPE);
+	return nullptr;
 }
 
 #define DECLARE_KNOWN_MONOCHROME_SETTING(key_, value_) {"cups-" key_, value_},
