--- a/gfx/layers/AnimationHelper.h
+++ b/gfx/layers/AnimationHelper.h
@@ -7,6 +7,7 @@
 #ifndef mozilla_layers_AnimationHelper_h
 #define mozilla_layers_AnimationHelper_h
 
+#include "X11UndefineNone.h"
 #include "mozilla/dom/Nullable.h"
 #include "mozilla/ComputedTimingFunction.h"  // for ComputedTimingFunction
 #include "mozilla/layers/AnimationStorageData.h"
@@ -15,7 +16,6 @@
 #include "mozilla/TimeStamp.h"                 // for TimeStamp
 #include "mozilla/TimingParams.h"
 #include "mozilla/Types.h"  // for SideBits
-#include "X11UndefineNone.h"
 #include <unordered_map>
 
 namespace mozilla {
--- a/widget/gtk/WindowSurfaceProvider.cpp	2022-01-14 06:26:17.000000000 +0800
+++ b/widget/gtk/WindowSurfaceProvider.cpp	2022-01-18 14:48:29.193204284 +0800
@@ -12,6 +12,7 @@
 #include "nsWindow.h"
 
 #ifdef MOZ_WAYLAND
+#  include "mozilla/WidgetUtilsGtk.h"
 #  include "mozilla/StaticPrefs_widget.h"
 #  include "WindowSurfaceWaylandMultiBuffer.h"
 #endif
--- a/widget/InProcessCompositorWidget.cpp	2022-01-24 15:35:11.376001012 +0800
+++ b/widget/InProcessCompositorWidget.cpp	2022-01-24 15:35:11.376001012 +0800
@@ -12,7 +12,7 @@
 
 // Platforms with no OOP compositor process support use
 // InProcessCompositorWidget by default.
-#if !defined(MOZ_WIDGET_SUPPORTS_OOP_COMPOSITING)
+#if !defined(MOZ_WIDGET_SUPPORTS_OOP_COMPOSITING) && !defined(MOZ_WAYLAND)
 /* static */
 RefPtr<CompositorWidget> CompositorWidget::CreateLocal(
     const CompositorWidgetInitData& aInitData,
--- a/widget/gtk/mozwayland/moz.build	2022-01-25 00:34:01.993828266 +0800
+++ b/widget/gtk/mozwayland/moz.build	2022-01-25 00:34:01.993828266 +0800
@@ -4,9 +4,6 @@
 # License, v. 2.0. If a copy of the MPL was not distributed with this
 # file, You can obtain one at http://mozilla.org/MPL/2.0/.
 
-SOURCES += [
-    "mozwayland.c",
-]
 EXPORTS.mozilla.widget += [
     "mozwayland.h",
 ]
--- a/widget/gtk/moz.build.orig	2022-01-25 11:04:49.650420160 +0800
+++ b/widget/gtk/moz.build	2022-01-25 11:05:04.417401668 +0800
@@ -23,6 +23,7 @@
     DIRS += ["mozgtk"]
 
 if CONFIG["MOZ_WAYLAND"]:
+    OS_LIBS += ["wayland-client", "wayland-egl"]
     DIRS += ["wayland", "mozwayland"]
 
 EXPORTS += [
