--- a/python/mach/mach/site.py	2022-04-12 06:05:07.000000000 +0800
+++ b/python/mach/mach/site.py	2022-04-12 06:05:07.000000000 +0800
@@ -777,7 +777,7 @@
         sysconfig_paths = sysconfig.get_paths(scheme)
         data_path = Path(sysconfig_paths["data"])
         path = Path(sysconfig_paths[sysconfig_path])
-        relative_path = path.relative_to(data_path)
+        relative_path = path.resolve().relative_to(data_path)
 
         # Path to virtualenv's "site-packages" directory for provided sysconfig path
         return os.path.normpath(Path(self.prefix) / relative_path)
