--- a/security/sandbox/linux/SandboxFilter.cpp
+++ b/security/sandbox/linux/SandboxFilter.cpp
@@ -1694,13 +1694,13 @@
         return Trap(OpenTrap, mFiles);
 
       case __NR_brk:
+      case __NR_sched_setscheduler:
         return Allow();
       case __NR_sched_get_priority_min:
       case __NR_sched_get_priority_max:
         return Allow();
       case __NR_sched_getparam:
-      case __NR_sched_getscheduler:
-      case __NR_sched_setscheduler: {
+      case __NR_sched_getscheduler: {
         Arg<pid_t> pid(0);
         return If(pid == 0, Allow()).Else(Trap(SchedTrap, nullptr));
       }
