--- media/ffvpx/config_unix64.h	2022-01-20 06:44:27.000000000 +0800
+++ media/ffvpx/config_unix64.h	2022-01-20 06:44:27.000000000 +0800
@@ -216,7 +216,7 @@
 #define HAVE_ES2_GL_H 0
 #define HAVE_GSM_H 0
 #define HAVE_IO_H 0
-#define HAVE_LINUX_PERF_EVENT_H 0
+#define HAVE_LINUX_PERF_EVENT_H 1
 #define HAVE_MACHINE_IOCTL_BT848_H 0
 #define HAVE_MACHINE_IOCTL_METEOR_H 0
 #define HAVE_OPENCV2_CORE_CORE_C_H 0
