--- a/c10/macros/Macros.h	2020-06-01 19:46:19.212517401 +0800
+++ b/c10/macros/Macros.h	2020-06-01 19:46:19.212517401 +0800
@@ -383,8 +383,8 @@
     __assert_fail(
         const char* assertion,
         const char* file,
-        unsigned int line,
-        const char* function) noexcept __attribute__((__noreturn__));
+        int line,
+        const char* function);
 
 #if (defined(__HIP_ARCH__) || defined(__HIP__)) && \
     !defined(TORCH_DISABLE_GPU_ASSERTS)
@@ -516,7 +516,7 @@
     (TARGET_IPHONE_SIMULATOR || TARGET_OS_SIMULATOR || TARGET_OS_IPHONE)
 #define HAS_DEMANGLE 0
 #else
-#define HAS_DEMANGLE 1
+#define HAS_DEMANGLE 0
 #endif
 #endif // HAS_DEMANGLE
 
