--- src/gallium/frontends/nine/threadpool.h.orig	2015-05-07 14:10:53.443337212 +0200
+++ src/gallium/frontends/nine/threadpool.h	2015-05-07 14:11:04.210307653 +0200
@@ -24,6 +24,8 @@
 #ifndef _THREADPOOL_H_
 #define _THREADPOOL_H_
 
+#include <pthread.h>
+
 #define MAXTHREADS 1
 
 struct threadpool {
--- src/util/rand_xor.c.orig	2020-10-03 12:27:48.489024729 +0200
+++ src/util/rand_xor.c	2020-10-03 12:31:05.927113521 +0200
@@ -28,6 +28,7 @@
 #if defined(HAVE_GETRANDOM)
 #include <sys/random.h>
 #endif
+#include <sys/types.h>		/* size_t, ssize_t */
 #include <unistd.h>
 #include <fcntl.h>
 #endif
--- meson.build	2023-08-20 12:33:35.535852780 +0800
+++ meson.build	2023-08-20 12:33:35.535852780 +0800
@@ -1626,7 +1626,7 @@
   pre_args += '-DHAVE_LIBUDEV'
 endif
 
-llvm_modules = ['bitwriter', 'engine', 'mcdisassembler', 'mcjit', 'core', 'executionengine', 'scalaropts', 'transformutils', 'instcombine']
+llvm_modules = ['bitwriter', 'engine', 'mcdisassembler', 'mcjit', 'core', 'executionengine', 'scalaropts', 'transformutils', 'instcombine', 'passes', 'ipo', 'orcjit']
 llvm_optional_modules = ['coroutines']
 if with_amd_vk or with_gallium_radeonsi or with_gallium_r600
   llvm_modules += ['amdgpu', 'bitreader', 'ipo']
