diff --git ./include/c11/threads_posix.h ./include/c11/threads_posix.h
index 45cb6075e6..1a2ea1a450 100644
--- ./src/c11/impl/threads_posix.c	2023-02-23 10:36:49.000000000 +0800
+++ ./src/c11/impl/threads_posix.c	2023-02-23 10:36:49.000000000 +0800
@@ -259,10 +259,24 @@
     if (!pack) return thrd_nomem;
     pack->func = func;
     pack->arg = arg;
-    if (pthread_create(thr, NULL, impl_thrd_routine, pack) != 0) {
+#ifdef __GLIBC__
+    pthread_attr_t *attrp = NULL;
+#else
+    pthread_attr_t attr = { 0 };
+    pthread_attr_init(&attr);
+    pthread_attr_setstacksize(&attr, 8388608);
+    pthread_attr_t *attrp = &attr;
+#endif
+    if (pthread_create(thr, attrp, impl_thrd_routine, pack) != 0) {
+#ifndef __GLIBC__
+        pthread_attr_destroy(&attr);
+#endif
         free(pack);
         return thrd_error;
     }
+#ifndef __GLIBC__
+    pthread_attr_destroy(&attr);
+#endif
     return thrd_success;
 }
 
