--- src/gallium/winsys/svga/drm/vmw_screen.h.orig	2015-05-07 14:02:28.033079796 +0200
+++ src/gallium/winsys/svga/drm/vmw_screen.h	2015-05-07 14:02:48.832054666 +0200
@@ -35,6 +35,8 @@
 #define VMW_SCREEN_H_
 
 
+#include <sys/types.h> 	/* dev_t */
+
 #include "pipe/p_compiler.h"
 #include "pipe/p_state.h"
 
--- src/gallium/frontends/nine/threadpool.h.orig	2015-05-07 14:10:53.443337212 +0200
+++ src/gallium/frontends/nine/threadpool.h	2015-05-07 14:11:04.210307653 +0200
@@ -24,6 +24,8 @@
 #ifndef _THREADPOOL_H_
 #define _THREADPOOL_H_
 
+#include <pthread.h>
+
 #define MAXTHREADS 1
 
 struct threadpool {
--- src/util/rand_xor.c.orig	2020-10-03 12:27:48.489024729 +0200
+++ src/util/rand_xor.c	2020-10-03 12:31:05.927113521 +0200
@@ -28,6 +28,7 @@
 #if defined(HAVE_GETRANDOM)
 #include <sys/random.h>
 #endif
+#include <sys/types.h>		/* size_t, ssize_t */
 #include <unistd.h>
 #include <fcntl.h>
 #endif
