--- cairo-1.16.0/configure
+++ cairo-1.16.0/configure
@@ -32219,7 +32219,7 @@
 fi
   if test "x$test_pdf" = "xyes"; then
 
-$as_echo "#define CAIRO_CAN_TEST_PDF_SURFACE 1" >>confdefs.h
+test_pdf=no
 
   else
     { $as_echo "$as_me:${as_lineno-$LINENO}: WARNING: PDF backend will not be tested since poppler >= $POPPLER_VERSION_REQUIRED is not available" >&5
@@ -32227,6 +32227,8 @@
   fi
 fi
 
+test_pdf=no
+any2ppm_pdf=no
  if test "x$test_pdf" = "xyes"; then
   CAIRO_CAN_TEST_PDF_SURFACE_TRUE=
   CAIRO_CAN_TEST_PDF_SURFACE_FALSE='#'
