--- a/hotspot/src/os/linux/vm/os_linux.cpp	2020-05-23 16:01:21.063604191 +0800
+++ b/hotspot/src/os/linux/vm/os_linux.cpp	2020-05-23 16:01:21.063604191 +0800
@@ -803,7 +803,7 @@
   // processors with hyperthreading technology.
   static int counter = 0;
   int pid = os::current_process_id();
-  alloca(((pid ^ counter++) & 7) * 128);
+  //alloca(((pid ^ counter++) & 7) * 128);
 
   ThreadLocalStorage::set_thread(thread);
 
