--- a/tools/cpp/unix_cc_configure.bzl	1980-01-01 00:00:00.000000000 +0800
+++ b/tools/cpp/unix_cc_configure.bzl	1980-01-01 00:00:00.000000000 +0800
@@ -348,7 +348,7 @@
         "",
         False,
     ), ":")
-    supports_gold_linker = _is_gold_supported(repository_ctx, cc)
+    supports_gold_linker = False
     cc_path = repository_ctx.path(cc)
     if not str(cc_path).startswith(str(repository_ctx.path(".")) + "/"):
         # cc is outside the repository, set -B
