--- sway/sway/main.c	2018-04-22 08:26:08.518588723 +0000
+++ sway/sway/main.c	2018-04-22 08:26:08.518588723 +0000
@@ -375,9 +375,6 @@
 	detect_proprietary();
 	detect_raspi();
 
-#ifdef __linux__
-	drop_permissions(suid);
-#endif
 	// handle SIGTERM signals
 	signal(SIGTERM, sig_handler);
 
@@ -414,6 +411,9 @@
 	// TODO: consume config->cmd_queue
 	config->active = true;
 
+#ifdef __linux__
+	drop_permissions(suid);
+#endif
 	if (!terminate_request) {
 		server_run(&server);
 	}
