diff --git a/arch/x86_64/reloc.h b/arch/x86_64/reloc.h
index fac0c0ae..fb6be3ea 100644
--- a/arch/x86_64/reloc.h
+++ b/arch/x86_64/reloc.h
@@ -10,6 +10,7 @@
 #define REL_DTPOFF      R_X86_64_DTPOFF64
 #define REL_TPOFF       R_X86_64_TPOFF64
 #define REL_TLSDESC     R_X86_64_TLSDESC
+#define REL_IRELATIVE   R_X86_64_IRELATIVE
 
 #define CRTJMP(pc,sp) __asm__ __volatile__( \
 	"mov %1,%%rsp ; jmp *%0" : : "r"(pc), "r"(sp) : "memory" )
diff --git a/ldso/dynlink.c b/ldso/dynlink.c
index fd09ca69..659cb113 100644
--- a/ldso/dynlink.c
+++ b/ldso/dynlink.c
@@ -507,6 +507,10 @@ static void do_relocs(struct dso *dso, size_t *rel, size_t rel_size, size_t stri
 			reloc_addr[1] = tmp;
 #endif
 			break;
+		case REL_IRELATIVE:
+			*reloc_addr = (size_t)base + addend;
+			*reloc_addr = ((size_t(*)(void))*reloc_addr)();
+			break;
 		default:
 			error("Error relocating %s: unsupported relocation type %d",
 				dso->name, type);
diff --git a/src/internal/dynlink.h b/src/internal/dynlink.h
index 830354eb..c67bc35e 100644
--- a/src/internal/dynlink.h
+++ b/src/internal/dynlink.h
@@ -41,6 +41,7 @@ enum {
 	REL_TPOFF,
 	REL_TPOFF_NEG,
 	REL_TLSDESC,
+	REL_IRELATIVE,
 	REL_FUNCDESC,
 	REL_FUNCDESC_VAL,
 };
