--- a/libgcompat/resolv.c	2022-10-30 05:39:20.000000000 +0800
+++ b/libgcompat/resolv.c	2022-10-30 05:39:20.000000000 +0800
@@ -12,6 +12,9 @@
 
 #include "alias.h" /* weak_alias */
 
+#include <fcntl.h>
+#include <stdarg.h>
+int __res_init() { return res_init(); }
 int __res_ninit(res_state statp)
 {
 	int rc;
--- a/libgcompat/execinfo.c	2022-10-30 05:39:20.000000000 +0800
+++ b/libgcompat/execinfo.c	2022-10-30 05:39:20.000000000 +0800
@@ -24,6 +24,7 @@
  */
 int backtrace(void **array, int size)
 {
+	return 0;
 	get_frame_level(array, size, 0);
 	get_frame_level(array, size, 1);
 	get_frame_level(array, size, 2);
