--- a/include/string.h	2022-04-08 01:12:40.000000000 +0800
+++ b/include/string.h	2022-04-08 01:12:40.000000000 +0800
@@ -88,7 +88,7 @@
 #endif
 
 #ifdef _GNU_SOURCE
-#define	strdupa(x)	strcpy(alloca(strlen(x)+1),x)
+#define	strdupa(x)	strcpy((char*)(alloca(strlen(x)+1)),x)
 int strverscmp (const char *, const char *);
 char *strchrnul(const char *, int);
 char *strcasestr(const char *, const char *);
