--- a/gtk/meson.build	2021-06-16 04:58:25.000000000 +0800
+++ b/gtk/meson.build	2021-06-16 04:58:25.000000000 +0800
@@ -992,15 +992,13 @@
     gtk_gir_public_headers,
     gtk_deprecated_headers,
     gtk_sources,
-    a11y_headers,
-    a11y_sources,
     gtktypebuiltins_h,
     gtkversion,
   ]
 
-  gtk_gir_includes = [gdk_gir[0], 'Atk-1.0']
-  gtk_gir_extra_args = gir_args + ['-DGTK_COMPILATION',
-                                   '--c-include=gtk/gtk-a11y.h']
+  gtk_gir_includes = [gdk_gir[0]]
+  gtk_gir_extra_args = gir_args + ['-DGTK_COMPILATION']
+
   if x11_enabled
     gtk_gir_includes += ['xlib-2.0']
     gtk_gir_extra_args += ['--c-include=gtk/gtkx.h']
