--- a/modules/printbackends/file/gtkprintbackendfile.c	2021-02-25 03:13:20.000000000 +0800
+++ b/modules/printbackends/file/gtkprintbackendfile.c	2021-02-25 03:13:20.000000000 +0800
@@ -30,7 +30,6 @@
 #include <cairo.h>
 #include <cairo-pdf.h>
 #include <cairo-ps.h>
-#include <cairo-svg.h>
 
 #include <glib/gi18n-lib.h>
 
@@ -340,8 +339,6 @@
 {
   cairo_surface_t *surface;
   OutputFormat format;
-  const cairo_svg_version_t *versions;
-  int num_versions = 0;
 
   format = format_from_settings (settings);
 
@@ -354,12 +351,6 @@
       case FORMAT_PS:
         surface = cairo_ps_surface_create_for_stream (_cairo_write, cache_io, width, height);
         break;
-      case FORMAT_SVG:
-        surface = cairo_svg_surface_create_for_stream (_cairo_write, cache_io, width, height);
-        cairo_svg_get_versions (&versions, &num_versions);
-        if (num_versions > 0)
-          cairo_svg_surface_restrict_to_version (surface, versions[num_versions - 1]);
-        break;
     }
 
   cairo_surface_set_fallback_resolution (surface,
