--- ppsspp/SDL/SDLGLGraphicsContext.cpp	2020-04-16 23:29:43.482342187 +0800
+++ ppsspp/SDL/SDLGLGraphicsContext.cpp	2020-04-16 23:29:43.482342187 +0800
@@ -87,17 +87,9 @@
 	SDL_VERSION(&sysInfo.version);
 	if (!SDL_GetWindowWMInfo(window, &sysInfo)) {
 		printf("ERROR: Unable to retrieve native window handle\n");
-		g_Display = (EGLNativeDisplayType)XOpenDisplay(nullptr);
-		g_XDisplayOpen = g_Display != nullptr;
-		if (!g_XDisplayOpen)
-			EGL_ERROR("Unable to get display!", false);
-		g_Window = (EGLNativeWindowType)nullptr;
+		EGL_ERROR("Unable to get display!", false);
 	} else {
 		switch (sysInfo.subsystem) {
-		case SDL_SYSWM_X11:
-			g_Display = (EGLNativeDisplayType)sysInfo.info.x11.display;
-			g_Window = (EGLNativeWindowType)sysInfo.info.x11.window;
-			break;
 #if defined(SDL_VIDEO_DRIVER_DIRECTFB)
 		case SDL_SYSWM_DIRECTFB:
 			g_Display = (EGLNativeDisplayType)EGL_DEFAULT_DISPLAY;
@@ -119,12 +111,7 @@
 		}
 
 		if (!EGL_OpenInit()) {
-			// Let's try again with X11.
-			g_Display = (EGLNativeDisplayType)XOpenDisplay(nullptr);
-			g_XDisplayOpen = g_Display != nullptr;
-			if (!g_XDisplayOpen)
-				EGL_ERROR("Unable to get display!", false);
-			g_Window = (EGLNativeWindowType)nullptr;
+			EGL_ERROR("Unable to get display!", false);
 		}
 	}

@@ -268,10 +268,6 @@
 		g_eglDisplay = EGL_NO_DISPLAY;
 	}
 	if (g_Display != nullptr) {
-#if !defined(USING_FBDEV)
-		if (g_XDisplayOpen)
-			XCloseDisplay((Display *)g_Display);
-#endif
 		g_XDisplayOpen = false;
 		g_Display = nullptr;
 	}
