--- systemd-246/src/home/homework-luks.c	2020-07-31 03:02:36.000000000 +0800
+++ systemd-246/src/home/homework-luks.c	2020-07-31 03:02:36.000000000 +0800
@@ -34,6 +34,7 @@
 #include "stat-util.h"
 #include "strv.h"
 #include "tmpfile-util.h"
+#include "missing_stdlib.h"
 
 /* Round down to the nearest 1K size. Note that Linux generally handles block devices with 512 blocks only,
  * but actually doesn't accept uneven numbers in many cases. To avoid any confusion around this we'll
--- systemd-246/src/home/homed-manager.c	2020-08-22 23:50:41.812026674 +0800
+++ systemd-246/src/home/homed-manager.c	2020-08-22 23:50:41.812026674 +0800
@@ -44,6 +44,7 @@
 #include "user-record-util.h"
 #include "user-record.h"
 #include "user-util.h"
+#include "missing_stdlib.h"
 
 /* Where to look for private/public keys that are used to sign the user records. We are not using
  * CONF_PATHS_NULSTR() here since we want to insert /var/lib/systemd/home/ in the middle. And we insert that
