--- systemd-245/src/home/user-record-sign.c	2020-03-06 20:38:52.000000000 +0800
+++ systemd-245/src/home/user-record-sign.c	2020-03-06 20:38:52.000000000 +0800
@@ -53,14 +53,16 @@
                 return -EIO;
 
         /* Request signature size */
-        if (EVP_DigestSign(md_ctx, NULL, &signature_size, (uint8_t*) text, strlen(text)) <= 0)
+        if (EVP_DigestSignUpdate(md_ctx, NULL, &signature_size) <= 0)
                 return -EIO;
 
         signature = malloc(signature_size);
         if (!signature)
                 return -ENOMEM;
 
-        if (EVP_DigestSign(md_ctx, signature, &signature_size, (uint8_t*) text, strlen(text)) <= 0)
+        if (EVP_DigestSignUpdate(md_ctx, signature, &signature_size) <= 0)
+                return -EIO;
+        if (EVP_DigestSignFinal(md_ctx, (uint8_t*) text, strlen(text)) <= 0)
                 return -EIO;
 
         mf = open_memstream_unlocked(&key, &key_size);
@@ -148,7 +148,7 @@
                 if (EVP_DigestVerifyInit(md_ctx, NULL, NULL, NULL, public_key) <= 0)
                         return -EIO;
 
-                if (EVP_DigestVerify(md_ctx, signature, signature_size, (uint8_t*) text, strlen(text)) <= 0) {
+                if (EVP_DigestVerifyUpdate(md_ctx, signature, signature_size) <= 0 || EVP_DigestVerifyFinal(md_ctx, (uint8_t*) text, strlen(text)) <= 0) {
                         n_bad ++;
                         continue;
                 }
--- systemd-245/src/home/homed-manager.c	2020-03-06 20:38:52.000000000 +0800
+++ systemd-245/src/home/homed-manager.c	2020-03-06 20:38:52.000000000 +0800
@@ -1288,19 +1288,19 @@
                 m->private_key = NULL;
         }
 
-        ctx = EVP_PKEY_CTX_new_id(EVP_PKEY_ED25519, NULL);
+        ctx = EVP_PKEY_CTX_new_id(EVP_PKEY_DSA2, NULL);
         if (!ctx)
-                return log_error_errno(SYNTHETIC_ERRNO(EIO), "Failed to allocate Ed25519 key generation context.");
+                return log_error_errno(SYNTHETIC_ERRNO(EIO), "Failed to allocate DSA2 key generation context.");
 
         if (EVP_PKEY_keygen_init(ctx) <= 0)
-                return log_error_errno(SYNTHETIC_ERRNO(EIO), "Failed to initialize Ed25519 key generation context.");
+                return log_error_errno(SYNTHETIC_ERRNO(EIO), "Failed to initialize DSA2 key generation context.");
 
         log_info("Generating key pair for signing local user identity records.");
 
         if (EVP_PKEY_keygen(ctx, &m->private_key) <= 0)
-                return log_error_errno(SYNTHETIC_ERRNO(EIO), "Failed to generate Ed25519 key pair");
+                return log_error_errno(SYNTHETIC_ERRNO(EIO), "Failed to generate DSA2 key pair");
 
-        log_info("Successfully created Ed25519 key pair.");
+        log_info("Successfully created DSA2 key pair.");
 
         (void) mkdir_p("/var/lib/systemd/home", 0755);
 
