--- systemd-245/src/basic/user-util.c	2020-03-06 20:38:52.000000000 +0800
+++ systemd-245/src/basic/user-util.c	2020-03-06 20:38:52.000000000 +0800
@@ -913,8 +913,6 @@
 
         errno = 0;
         p = fgetpwent(stream);
-        if (!p && errno != ENOENT)
-                return errno_or_else(EIO);
 
         *pw = p;
         return !!p;
@@ -928,8 +926,6 @@
 
         errno = 0;
         s = fgetspent(stream);
-        if (!s && errno != ENOENT)
-                return errno_or_else(EIO);
 
         *sp = s;
         return !!s;
@@ -943,8 +939,6 @@
 
         errno = 0;
         g = fgetgrent(stream);
-        if (!g && errno != ENOENT)
-                return errno_or_else(EIO);
 
         *gr = g;
         return !!g;
@@ -959,8 +953,6 @@
 
         errno = 0;
         s = fgetsgent(stream);
-        if (!s && errno != ENOENT)
-                return errno_or_else(EIO);
 
         *sg = s;
         return !!s;
--- systemd-247/src/basic/alloc-util.h	2020-11-27 02:00:50.000000000 +0800
+++ systemd-247/src/basic/alloc-util.h	2020-11-27 02:00:50.000000000 +0800
@@ -96,7 +96,7 @@
         return malloc(size * need ?: 1);
 }
 
-#if !HAVE_REALLOCARRAY
+#if 0
 _alloc_(2, 3) static inline void *reallocarray(void *p, size_t need, size_t size) {
         if (size_multiply_overflow(size, need))
                 return NULL;
