--- systemd-240/src/login/logind-dbus.c
+++ systemd-240/src/login/logind-dbus.c
@@ -1739,11 +1739,7 @@
         }
 
         if (!multiple_sessions && !blocked && action) {
-                r = bus_verify_polkit_async(message, CAP_SYS_BOOT, action, NULL, interactive, UID_INVALID, &m->polkit_registry, error);
-                if (r < 0)
-                        return r;
-                if (r == 0)
-                        return 1; /* No authorization for now, but the async polkit stuff will call us again when it has it */
+					return 0;
         }
 
         return 0;
