--- poppler-0.64.0/utils/pdfsig.cc	2018-04-18 02:04:26.000000000 +0800
+++ poppler-0.64.0/utils/pdfsig.cc	2018-04-18 02:04:26.000000000 +0800
@@ -19,6 +19,7 @@
 #include <stdio.h>
 #include <stdlib.h>
 #include <stddef.h>
+#include <libgen.h>
 #include <string.h>
 #include <time.h>
 #include <hasht.h>
--- poppler-0.64.0/utils/pdfsig.cc	2018-05-28 01:34:51.401371713 +0800
+++ poppler-0.64.0/utils/pdfsig.cc	2018-05-28 01:34:51.401371713 +0800
@@ -105,7 +105,7 @@
     // since { is the magic character to replace things we need to put it twice where
     // we don't want it to be replaced
     GooString *format = GooString::format("{{0:s}}.sig{{1:{0:d}d}}", sigCountLength);
-    GooString *path = GooString::format(format->getCString(), basename(filename), sig_num);
+    GooString *path = GooString::format(format->getCString(), basename((char*)filename), sig_num);
     printf("Signature #%d (%u bytes) => %s\n", sig_num, signature->getLength(), path->getCString());
     std::ofstream outfile(path->getCString(), std::ofstream::binary);
     outfile.write(signature->getCString(), signature->getLength());
