--- a/cmake/CMakeLists.txt.orig	2023-04-02 21:29:32.083574207 +0800
+++ b/cmake/CMakeLists.txt	2023-04-03 00:35:20.296601126 +0800
@@ -808,6 +808,7 @@
 if (onnxruntime_PREFER_SYSTEM_LIB)
   find_package(Protobuf)
 endif()
+find_package(Protobuf REQUIRED)
 if (Protobuf_FOUND)
   message("Use protobuf from preinstalled system lib")
   if (onnxruntime_USE_FULL_PROTOBUF)
@@ -1605,7 +1605,6 @@
 
 # Flatbuffers
 if (onnxruntime_PREFER_SYSTEM_LIB)
-  find_package(Flatbuffers)
 endif()
 if (Flatbuffers_FOUND)
   message("Use flatbuffers from preinstalled system lib")
@@ -1608,7 +1610,7 @@
 endif()
 if (Flatbuffers_FOUND)
   message("Use flatbuffers from preinstalled system lib")
-  add_library(flatbuffers ALIAS flatbuffers::flatbuffers)
+  add_library(flatbuffers ALIAS flatbuffers::flatbuffers_shared)
 else()
   message("Use flatbuffers from submodule")
   # We do not need to build flatc for iOS or Android Cross Compile
