--- a/drivers/gpu/drm/nouveau/nouveau_drm.c
+++ b/drivers/gpu/drm/nouveau/nouveau_drm.c
@@ -89,6 +89,10 @@
 static int nouveau_runtime_pm = -1;
 module_param_named(runpm, nouveau_runtime_pm, int, 0400);
 
+MODULE_PARM_DESC(runpm_freeze_workaround, "disable (0),  enable (1), default 0");
+static int nouveau_runtime_pm_freeze_workaround = 0;
+module_param_named(runpm_freeze_workaround, nouveau_runtime_pm_freeze_workaround, int, 0400);
+
 static struct drm_driver driver_stub;
 static struct drm_driver driver_pci;
 static struct drm_driver driver_platform;
@@ -835,12 +839,17 @@
 		return -EBUSY;
 	}
 
+	drm_dev->switch_power_state = DRM_SWITCH_POWER_CHANGING;
 	nouveau_switcheroo_optimus_dsm();
 	ret = nouveau_do_suspend(drm_dev, true);
 	pci_save_state(pdev);
 	pci_disable_device(pdev);
 	pci_ignore_hotplug(pdev);
-	pci_set_power_state(pdev, PCI_D3cold);
+	if (nouveau_runtime_pm_freeze_workaround) {
+		__pci_complete_power_transition(pdev, PCI_D3cold);
+	} else {
+		pci_set_power_state(pdev, PCI_D3cold);
+	}
 	drm_dev->switch_power_state = DRM_SWITCH_POWER_DYNAMIC_OFF;
 	return ret;
 }
@@ -858,7 +867,12 @@
 		return -EBUSY;
 	}
 
-	pci_set_power_state(pdev, PCI_D0);
+	drm_dev->switch_power_state = DRM_SWITCH_POWER_CHANGING;
+	if (nouveau_runtime_pm_freeze_workaround) {
+		__pci_complete_power_transition(pdev, PCI_D0);
+	} else {
+		pci_set_power_state(pdev, PCI_D0);
+	}
 	pci_restore_state(pdev);
 	ret = pci_enable_device(pdev);
 	if (ret)
