--- a/drivers/net/wireless/intel/iwlwifi/iwl-modparams.h
+++ b/drivers/net/wireless/intel/iwlwifi/iwl-modparams.h
@@ -130,6 +130,8 @@
 	int amsdu_size;
 	bool fw_restart;
 	bool bt_coex_active;
+	int bt_coex_en_thresh;
+	int bt_coex_dis_thresh;
 	int led_mode;
 	bool power_save;
 	int power_level;
--- a/drivers/net/wireless/intel/iwlwifi/iwl-drv.c	2018-09-29 17:56:03.000000000 +0800
+++ b/drivers/net/wireless/intel/iwlwifi/iwl-drv.c	2018-09-29 17:56:03.000000000 +0800
@@ -1829,6 +1829,13 @@
 		   bool, 0444);
 MODULE_PARM_DESC(bt_coex_active, "enable wifi/bt co-exist (default: enable)");
 
+module_param_named(bt_coex_en_thresh, iwlwifi_mod_params.bt_coex_en_thresh,
+		   int, 0444);
+MODULE_PARM_DESC(bt_coex_active, "wifi/bt co-exist en thresh (default: 62)");
+module_param_named(bt_coex_dis_thresh, iwlwifi_mod_params.bt_coex_dis_thresh,
+		   int, 0444);
+MODULE_PARM_DESC(bt_coex_active, "wifi/bt co-exist dis thresh (default: 65)");
+
 module_param_named(led_mode, iwlwifi_mod_params.led_mode, int, 0444);
 MODULE_PARM_DESC(led_mode, "0=system default, "
 		"1=On(RF On)/Off(RF Off), 2=blinking, 3=Off (default: 0)");
--- a/drivers/net/wireless/intel/iwlwifi/mvm/coex.c	2018-09-29 17:56:03.000000000 +0800
+++ b/drivers/net/wireless/intel/iwlwifi/mvm/coex.c	2018-09-29 17:56:03.000000000 +0800
@@ -292,9 +292,9 @@
 
 	mvmvif->bf_data.last_bt_coex_event = rssi;
 	mvmvif->bf_data.bt_coex_max_thold =
-		enable ? -IWL_MVM_BT_COEX_EN_RED_TXP_THRESH : 0;
+		enable ? -iwlwifi_mod_params.bt_coex_en_thresh : 0;
 	mvmvif->bf_data.bt_coex_min_thold =
-		enable ? -IWL_MVM_BT_COEX_DIS_RED_TXP_THRESH : 0;
+		enable ? -iwlwifi_mod_params.bt_coex_dis_thresh : 0;
 }
 
 #define MVM_COEX_TCM_PERIOD (HZ * 10)
@@ -454,10 +454,10 @@
 	/* if the RSSI isn't valid, fake it is very low */
 	if (!ave_rssi)
 		ave_rssi = -100;
-	if (ave_rssi > -IWL_MVM_BT_COEX_EN_RED_TXP_THRESH) {
+	if (ave_rssi > -iwlwifi_mod_params.bt_coex_en_thresh) {
 		if (iwl_mvm_bt_coex_reduced_txp(mvm, mvmvif->ap_sta_id, true))
 			IWL_ERR(mvm, "Couldn't send BT_CONFIG cmd\n");
-	} else if (ave_rssi < -IWL_MVM_BT_COEX_DIS_RED_TXP_THRESH) {
+	} else if (ave_rssi > -iwlwifi_mod_params.bt_coex_dis_thresh) {
 		if (iwl_mvm_bt_coex_reduced_txp(mvm, mvmvif->ap_sta_id, false))
 			IWL_ERR(mvm, "Couldn't send BT_CONFIG cmd\n");
 	}
--- a/drivers/net/wireless/intel/iwlwifi/mvm/constants.h	2018-09-29 17:56:03.000000000 +0800
+++ b/drivers/net/wireless/intel/iwlwifi/mvm/constants.h	2018-09-29 17:56:03.000000000 +0800
@@ -94,8 +94,6 @@
 #define IWL_MVM_PS_SNOOZE_WINDOW		50
 #define IWL_MVM_WOWLAN_PS_SNOOZE_WINDOW		25
 #define IWL_MVM_LOWLAT_QUOTA_MIN_PERCENT	64
-#define IWL_MVM_BT_COEX_EN_RED_TXP_THRESH	62
-#define IWL_MVM_BT_COEX_DIS_RED_TXP_THRESH	65
 #define IWL_MVM_BT_COEX_SYNC2SCO		1
 #define IWL_MVM_BT_COEX_MPLUT			1
 #define IWL_MVM_BT_COEX_RRC			1
