--- dwm-6.1/dwm.c	2015-11-08 22:39:37.000000000 +0000
+++ dwm-6.1/dwm.c	2015-11-08 22:39:37.000000000 +0000
@@ -720,7 +720,10 @@
 		w = TEXTW(tags[i]);
 		drw_setscheme(drw, m->tagset[m->seltags] & 1 << i ? &scheme[SchemeSel] : &scheme[SchemeNorm]);
-		drw_text(drw, x, 0, w, bh, tags[i], urg & 1 << i);
+		drw_text(drw, x, 0, w, bh, tags[i], urg & 1 << i);
+		if (m == selmon && selmon->sel && selmon->sel->tags & 1 << i) {
+		    drw_rect(drw, x, bh-dx*2/3, w, dx*2/3, 1, 0, urg & 1 << i);
+		} else if (occ & 1 << i) {
-		drw_rect(drw, x + 1, 1, dx, dx, m == selmon && selmon->sel && selmon->sel->tags & 1 << i,
-		           occ & 1 << i, urg & 1 << i);
+		    drw_rect(drw, x + 1, 1, dx, dx, 0, 1, urg & 1 << i);
+               }
 		x += w;
 	}
