--- a/sip/cpp/sip_gridwxGridEvent.cpp	2020-11-22 03:10:38.000000000 +0800
+++ b/sip/cpp/sip_gridwxGridEvent.cpp	2020-11-22 03:10:38.000000000 +0800
@@ -32,8 +32,8 @@
      * this class.
      */
 protected:
-    int GetRow() SIP_OVERRIDE;
-    int GetCol() SIP_OVERRIDE;
+    int GetRow();
+    int GetCol();
      ::wxEvent* Clone() const SIP_OVERRIDE;
      ::wxEventCategory GetEventCategory() const SIP_OVERRIDE;
 
