--- ImHex/plugins/libimhex/source/ui/imgui_imhex_extensions.cpp	2022-01-10 14:52:29.875748172 +0800
+++ ImHex/plugins/libimhex/source/ui/imgui_imhex_extensions.cpp	2022-01-10 14:52:29.875748172 +0800
@@ -11,7 +11,7 @@
 
 #include <string>
 
-#include <imgui_impl_opengl3_loader.h>
+#include <GLFW/glfw3.h>
 
 namespace ImGui {
 
--- ImHex/source/window/window.cpp	2022-01-10 14:29:37.146517784 +0800
+++ ImHex/source/window/window.cpp	2022-01-10 14:29:37.146517784 +0800
@@ -759,8 +759,6 @@
         #endif
 
         glfwWindowHint(GLFW_CONTEXT_VERSION_MAJOR, 3);
-        glfwWindowHint(GLFW_CONTEXT_VERSION_MINOR, 2);
-        glfwWindowHint(GLFW_OPENGL_PROFILE, GLFW_OPENGL_CORE_PROFILE);
-        glfwWindowHint(GLFW_OPENGL_FORWARD_COMPAT, GL_TRUE);
-        glfwWindowHint(GLFW_VISIBLE, GLFW_FALSE);
+        glfwWindowHint(GLFW_CONTEXT_VERSION_MINOR, 0);
+        glfwWindowHint(GLFW_CONTEXT_CREATION_API, GLFW_EGL_CONTEXT_API);

         this->m_windowTitle = "ImHex";
         this->m_window = glfwCreateWindow(1280 * SharedData::globalScale, 720 * SharedData::globalScale, this->m_windowTitle.c_str(), nullptr, nullptr);
--- ImHex/source/init/splash_window.cpp	2022-01-10 14:29:37.146517784 +0800
+++ ImHex/source/init/splash_window.cpp	2022-01-10 14:29:37.146517784 +0800
@@ -155,8 +155,7 @@
         }
 
         glfwWindowHint(GLFW_CONTEXT_VERSION_MAJOR, 3);
-        glfwWindowHint(GLFW_CONTEXT_VERSION_MINOR, 2);
-        glfwWindowHint(GLFW_OPENGL_PROFILE, GLFW_OPENGL_CORE_PROFILE);
-        glfwWindowHint(GLFW_OPENGL_FORWARD_COMPAT, GL_TRUE);
+        glfwWindowHint(GLFW_CONTEXT_VERSION_MINOR, 0);
+        glfwWindowHint(GLFW_CONTEXT_CREATION_API, GLFW_EGL_CONTEXT_API);
 
         glfwWindowHint(GLFW_RESIZABLE, GLFW_FALSE);
         glfwWindowHint(GLFW_TRANSPARENT_FRAMEBUFFER, GLFW_TRUE);
--- ImHex-1.6.1/external/ImGui/CMakeLists.txt	2021-01-11 16:56:09.000000000 +0800
+++ ImHex-1.6.1/external/ImGui/CMakeLists.txt	2021-01-11 16:56:09.000000000 +0800
@@ -10,6 +10,7 @@
 
 if (UNIX)
     find_package(OpenGL REQUIRED)
+    add_compile_definitions(IMGUI_IMPL_OPENGL_ES2)
 endif ()
 
 add_library(imgui OBJECT
@@ -46,5 +47,5 @@
 if (WIN32)
     target_link_libraries(imgui PUBLIC Freetype::Freetype glfw3 opengl32.lib)
 elseif (UNIX)
-    target_link_libraries(imgui PUBLIC Freetype::Freetype glfw OpenGL::GL)
+    target_link_libraries(imgui PUBLIC Freetype::Freetype glfw OpenGL::OpenGL)
 endif()
