--- a/src/librustc_target/spec/mod.rs.orig     2019-05-23 21:05:45.173347119 +0200
+++ b/src/librustc_target/spec/mod.rs  2019-05-23 21:06:39.006681699 +0200
@@ -328,6 +328,7 @@ macro_rules! supported_targets {
 }
 
 supported_targets! {
+    ("x86_64-pc-linux-musl", x86_64_pc_linux_musl),
     ("x86_64-unknown-linux-gnu", x86_64_unknown_linux_gnu),
     ("x86_64-unknown-linux-gnux32", x86_64_unknown_linux_gnux32),
     ("i686-unknown-linux-gnu", i686_unknown_linux_gnu),
--- /dev/null
+++ b/src/librustc_target/spec/x86_64_pc_linux_musl.rs
@@ -0,0 +1,11 @@
+use crate::spec::TargetResult;
+
+pub fn target() -> TargetResult {
+    let mut base = super::x86_64_unknown_linux_musl::target()?;
+
+    base.llvm_target = "x86_64-pc-linux-musl".to_string();
+    base.target_vendor = "noname".to_string();
+    base.options.crt_static_default = false;
+
+    Ok(base)
+}
