diff --git a/p2p/base/port.cc b/p2p/base/port.cc
index a1b478d11..81aa0aadb 100644
--- a/third_party/webrtc/p2p/base/port.cc
+++ b/third_party/webrtc/p2p/base/port.cc
@@ -10,6 +10,8 @@
 
 #include "p2p/base/port.h"
 
+#include <math.h>
+
 #include <algorithm>
 #include <vector>
 
--- a/third_party/webrtc/modules/desktop_capture/linux/window_capturer_x11.cc.orig	2019-01-20 10:57:09.166224517 +0000
+++ b/third_party/webrtc/modules/desktop_capture/linux/window_capturer_x11.cc	2019-01-20 06:01:05.901545307 +0000
@@ -14,6 +14,7 @@
 #include <X11/extensions/Xcomposite.h>
 #include <X11/extensions/Xrender.h>
 
+#include <string.h>
 #include <memory>
 #include <string>
 #include <utility>
