--- tools/gn/config_values_generator.cc	2016-12-09 05:04:35.000000000 +0000
+++ tools/gn/config_values_generator.cc	2016-12-09 05:04:35.000000000 +0000
@@ -114,7 +114,7 @@
     config_values_->set_precompiled_source(
         input_dir_.ResolveRelativeFile(
             *precompiled_source_value, err_,
-            scope_->settings()->build_settings()->root_path_utf8()));
+            "/"));
     if (err_->has_error())
       return;
   }
