--- gpu/command_buffer/service/shared_image_factory.cc	2020-12-18 07:08:48.000000000 +0800
+++ gpu/command_buffer/service/shared_image_factory.cc	2020-12-18 07:08:48.000000000 +0800
@@ -41,7 +41,7 @@
 #include "gpu/vulkan/vulkan_device_queue.h"
 #elif defined(OS_MAC)
 #include "gpu/command_buffer/service/shared_image_backing_factory_iosurface.h"
-#elif BUILDFLAG(IS_ASH)
+#else
 #include "gpu/command_buffer/service/shared_image_backing_factory_ozone.h"
 #endif
 
@@ -146,17 +146,11 @@
   // OSX
   DCHECK(gr_context_type_ == GrContextType::kGL ||
          gr_context_type_ == GrContextType::kMetal);
-#elif BUILDFLAG(IS_ASH)
+#else
   if (gr_context_type_ == GrContextType::kVulkan) {
     interop_backing_factory_ =
         std::make_unique<SharedImageBackingFactoryOzone>(context_state);
   }
-#else
-  // Others
-  if (gr_context_type_ == GrContextType::kVulkan) {
-    LOG(ERROR) << "ERROR: gr_context_type_ is GrContextType::kVulkan and "
-                  "interop_backing_factory_ is not set";
-  }
 #endif
   if (enable_wrapped_sk_image && context_state) {
     wrapped_sk_image_factory_ =
